azkaban-common/src/main/java/azkaban/utils/Emailer.java

```diff
@@ -146,6 +153,55 @@ public void alertOnSuccess(final ExecutableFlow flow) {
     sendEmail(message, mailCreated, "success email message for execution" + flow.getExecutionId());
   }
 
+  /**
+   * Sends as many emails as there are unique combinations of:
+   *
+   * [mail creator] x [failure email address list]
+   *
+   * Executions with the same combo are grouped into a single message.
+   */
+  @Override
+  public void alertOnFailedUpdate(final Executor executor, List<ExecutableFlow> flows,
+      final ExecutorManagerException updateException) {
+
+    flows = flows.stream()
+        .filter(flow -> flow.getExecutionOptions() != null)
+        .filter(flow -> CollectionUtils.isNotEmpty(flow.getExecutionOptions().getFailureEmails()))
+        .collect(Collectors.toList());
+
+    // group by mail creator in case some flows use different creators
+    final ImmutableListMultimap<String, ExecutableFlow> creatorsToFlows = Multimaps
```

Comment:
It seems we only have one mailCreator per flow today. Do you think it's necessary to do the grouping here?

Reply:
This is needed to be able to send a single e-mail for any executions that have the same mail creator & recipients. Flows that have different mail creators can not be combined into a single message.

Other option would be to send separate e-mail for every execution, but I think it's definitely nicer to group them.

Reply:
Ok got it, thanks.